ImpPager.class.nut

```diff
@@ -36,78 +47,152 @@ class ImpPager {
-                    next(false);
-                    return;
-                }
-                // Save SPI Flash address in the message metadata
-                dataPoint.metadata <- {"addr" : addr};
-                _resendExistingMessage(dataPoint);
-                // Don't do any further scanning until we get an ACK for already sent message
-                next(false);
-            }.bindenv(this),
-            function() {
-                _log_debug("Finished processing all pending messages");
-            }.bindenv(this)
-        );
+        if(typeof(_next) == "function"){  // We were already in the middle of a read - continue from where we started.
+            _next(true);
+            _next = null;
+        } else {
+            _log_debug("Start processing pending messages...");
+            _spiFlashLogger.read(
+                function(dataPoint, addr, next) {
+                    _log_debug("Reading from SPI Flash. ID: " + dataPoint.id + " at addr: " + addr);
+
+                    // There's no point of retrying to send pending messages when disconnected
+                    if (!_connectionManager.isConnected()) {
+                        _log_debug("No connection, abort SPI Flash scanning...");
+                        // Abort scanning
+                        next(false);
+                        return;
+                    }
+
+                    if(time() == RTC_INVALID_TIME){ // If time is invalid, we aren't ready to resend any data just yet...
+                        _log_debug("time() was invalid, abort SPI Flash scanning...");
+                        next(false);
+                        return;
+                    }
+
+                    // Save SPI Flash address in the message metadata
+                    if(!("metadata" in dataPoint)) dataPoint.metadata <- {}
+                    dataPoint.metadata.addr <- addr;
+
+                    if("rtc" in dataPoint.metadata && dataPoint.metadata.rtc == false){
+                      if(_lastTS == null){
+                        _lastTS = [hardware.millis(), time()] //With these two datapoints, we can now re-establish all of our timestamps
+                        _log_debug("Discovered most recent datapoint saved to SPIFlash without RTC - " + dataPoint.id + " attempting to rebuild timestamps with ms = " + _lastTS[0] + " and time = " + _lastTS[1])
+                      }
+
+                      _log_debug("Found log without RTC. ID=" + dataPoint.id + " and ts=" +dataPoint.ts)
+
+                      if("boot" in dataPoint.metadata && dataPoint.metadata.boot == _bootNumber){
+                        local deltaTMillis = _lastTS[0] - dataPoint.ts
+                        local deltaTSeconds = math.floor(deltaTMillis+500).tointeger()/1000 //Round to nearest second, but use this int value for updating _lastTS to keep things consistent
+                        dataPoint.ts = _lastTS[1] - deltaTSeconds  //All integer math, so no need to worry about decimal points
```

Comment:
And as I mentioned in the Bullwinkle PR, I'm not fully convinced yet that we should modify .ts from outside the Bullwinkle.